proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,489 @@
+
+```kotlin
+package com.data.instances
+
+import com.data.Repository
+import com.domain.User
+import com.domain.Group
+
+extension class GroupRepository<A>(with val repoA: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): List<Group<A>> {
+    return listOf(Group(repoA.loadAll()))
+  }
+
+  override fun loadById(id: Int): Group<A>? {
+    return Group(repoA.loadById(id))
+  }
+}
+```
+
+The above extension provides evidence of a `Repository<Group<A>>` as long as there is a `Repository<A>` in scope. Call site would be like:
+
+```kotlin
+fun fetchGroup<A>(with repo: GroupRepository<A>) = repo.loadAll()
+
+fun main() {
+  fetchGroup<User>() // Succeeds! There's evidence of Repository<Group<A>> and Repository<User> provided in scope.
+  fetchGroup<Coin>() // Fails! There's evidence of Repository<Group<A>> but no evidence of `Repository<Coin>` available.
+}
+```
+
+We believe the encoding proposed above fits nicely with Kotlin's philosophy of extensions and will reduce the boilerplate compared to other langs that also support compile time dependency resolution such as Scala where this is done via implicits.
+
+## Language changes
+
+* Add `with` to require evidence of extensions in both function and class/object declarations.
+* Add `extension` to provide instance evidence for a given interface.
+
+Usage of these language changes are demonstrated by the previous and below examples:
+
+#### Class constraint
+
+```kotlin
+extension class GroupRepository<A>(with R: Repository<A>) : Repository<Group<A>> {
+  /* ... */
+}
+```
+
+#### Function constraint
+
+```kotlin
+fun <A> fetch(id: String, with R: Repository<A>): A = loadById(id) // function position using parameter "R"
+```
+
+#### Extension evidence using an Object
+
+```kotlin
+extension object UserRepository: Repository<User> {
+  override fun loadAll(): List<User> {
+    return listOf(User(25, "Bob"))
+  }
+
+  override fun loadById(id: Int): User? {
+    return if (id == 25) {
+      User(25, "Bob")
+    } else {
+      null
+    }
+  }
+}
+```
+
+#### Extension evidence using a Class
+
+```kotlin
+extension class UserRepository: Repository<User> {
+  override fun loadAll(): List<User> {
+    return listOf(User(25, "Bob"))
+  }
+
+  override fun loadById(id: Int): User? {
+    return if (id == 25) {
+      User(25, "Bob")
+    } else {
+      null
+    }
+  }
+}
+```
+
+## Extension resolution order
+
+Classical interfaces only permit their implementation at the site of a type definition. Compiler extension resolution pattern typically relax this rule and **allow extension evidences be declared outside of the type definition**. When relaxing this rule it is important to preserve the coherency we take for granted with classical interfaces.
+
+For those reasons constraint interfaces must be declared in one of the following places (in strict resolution order):
```

Comment:
`declared` is not the correct word because an argument in 1 is not a declaration.

Reply:
Fixed in 67c8f3e.